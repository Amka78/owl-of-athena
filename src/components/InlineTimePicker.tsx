import React, { Component } from "react";
import {
    Text,
    TouchableOpacity,
    View,
    StyleSheet,
    ViewStyle,
    TextStyle
} from "react-native";
import { Ionicons } from "@expo/vector-icons";

type TimePickerStyle = {
    textColor?: string;
    backgroundColor?: string;
    containerBackgroudColor?: string;
    activeColor?: string;
    borderColor?: string;
    fontSize?: number;
    borderRadius?: number;
    iconSize?: number;
};

const TimePickerDefaultStyle: TimePickerStyle = {
    textColor: "#ddd",
    backgroundColor: "#777",
    containerBackgroudColor: "#555",
    activeColor: "#000",
    borderColor: "#555",
    fontSize: 35,
    borderRadius: 4,
    iconSize: 35
};

type TimePickerMode = "full" | "minute";

type InlineTimePickerProps = {
    style?: TimePickerStyle;
    initialTime?: Date | { hours: number; minutes: number; seconds: number };
    onChangeTime?: (hours?: number, minutes?: number, seconds?: number) => void;
    mode24hours?: boolean;
    mode?: TimePickerMode;
};

type InlineTimePickerState = {
    hours: number;
    minutes: number;
    seconds: number;
    meridian: "AM" | "PM";
    updating?: Text;
};

export class InlineTimePicker extends Component<
    InlineTimePickerProps,
    InlineTimePickerState
> {
    private style: TimePickerStyle;

    private hoursText?: Text;

    private minutesText?: Text;

    private secondsText?: Text;

    private interval?: NodeJS.Timeout;

    private mode?: TimePickerMode;
    constructor(props: InlineTimePickerProps) {
        super(props);

        console.debug("InlineTimePickerProps:", props);
        this.state = {
            hours: 12,
            minutes: 0,
            seconds: 0,
            meridian: "AM",
            updating: undefined
        };
        this.hoursText = undefined;
        this.minutesText = undefined;
        this.secondsText = undefined;
        this.interval = undefined;

        this.mode = "full";
        if (this.props.mode) {
            this.mode = this.props.mode!;
        }

        this.style = TimePickerDefaultStyle;
        if (this.props.style) {
            this.style = this.props.style!;

            this.setStyle();
        }
    }

    public async componentDidMount(): Promise<void> {
        let initialDate = new Date();

        if (this.props.initialTime !== undefined) {
            if ("hours" in this.props.initialTime) {
                initialDate = new Date(
                    initialDate.getFullYear(),
                    initialDate.getMonth(),
                    initialDate.getDate(),
                    this.props.initialTime!.hours,
                    this.props.initialTime!.minutes,
                    this.props.initialTime!.seconds
                );
            } else {
                initialDate = this.props.initialTime;
            }
        }

        const currentHours = initialDate.getHours();
        const calculatedHours =
            this.props.mode24hours === true
                ? currentHours
                : currentHours >= 12
                ? currentHours - 12
                : currentHours;
        this.setState({
            hours: calculatedHours,
            minutes: initialDate.getMinutes(),
            seconds: initialDate.getSeconds(),
            meridian: currentHours > 12 ? "PM" : "AM"
        });
    }

    public async componentDidUpdate(
        _prevProps: InlineTimePickerProps,
        prevState: InlineTimePickerState
    ): Promise<void> {
        if (this.state !== prevState) this.invokeOnChangeTime();
    }

    public render(): JSX.Element {
        return (
            <View>
                <View style={[styles.container, this.getContainerColor()]}>
                    <View style={styles.timeContainer}>
                        {this.state.updating === undefined &&
                            !this.props.mode24hours && (
                                <TouchableOpacity
                                    style={{
                                        justifyContent: "center",
                                        marginRight: 10
                                    }}
                                    onPress={(): void => {
                                        this.setState({
                                            meridian:
                                                this.state.meridian === "AM"
                                                    ? "PM"
                                                    : "AM"
                                        });
                                    }}
                                >
                                    <Text
                                        style={[
                                            styles.text,
                                            this.getTextStyle()
                                        ]}
                                    >
                                        {this.state.meridian}
                                    </Text>
                                </TouchableOpacity>
                            )}
                        <TouchableOpacity
                            onPress={(): void => this.update(this.hoursText!)}
                        >
                            <Text
                                style={[styles.text, this.getTextStyle()]}
                                ref={(c): void => {
                                    this.hoursText = c!;
                                }}
                            >
                                {this.state.hours < 10 ? "0" : ""}
                                {this.state.hours}
                            </Text>
                        </TouchableOpacity>
                        <Text style={[styles.colon, this.getTextColor()]}>
                            {":"}
                        </Text>
                        <TouchableOpacity
                            onPress={(): void => this.update(this.minutesText!)}
                        >
                            <Text
                                style={[styles.text, this.getTextStyle()]}
                                ref={(c): void => {
                                    this.minutesText = c!;
                                }}
                            >
                                {this.state.minutes < 10 ? "0" : ""}
                                {this.state.minutes}
                            </Text>
                        </TouchableOpacity>
                        {this.mode === "full" && (
                            <Text style={[styles.colon, this.getTextColor()]}>
                                {":"}
                            </Text>
                        )}
                        {this.mode === "full" && (
                            <TouchableOpacity
                                onPress={(): void =>
                                    this.update(this.secondsText!)
                                }
                            >
                                <Text
                                    style={[styles.text, this.getTextStyle()]}
                                    ref={(c): void => {
                                        this.secondsText = c!;
                                    }}
                                >
                                    {this.state.seconds < 10 ? "0" : ""}
                                    {this.state.seconds}
                                </Text>
                            </TouchableOpacity>
                        )}
                    </View>
                    {this.state.updating !== undefined && (
                        <View style={styles.center}>
                            <TouchableOpacity
                                style={[
                                    styles.text,
                                    this.getTextStyle(),
                                    styles.increment
                                ]}
                                onPress={(): void => this.increment(1)}
                                onLongPress={(): void => this.longIncrement(10)}
                                onPressOut={(): void => this.longIncrement()}
                            >
                                <Ionicons
                                    name={"md-add"}
                                    size={this.style.iconSize}
                                    color={this.style.textColor}
                                />
                            </TouchableOpacity>
                            <TouchableOpacity
                                style={[
                                    styles.text,
                                    this.getTextStyle(),
                                    styles.increment
                                ]}
                                onPress={(): void => this.increment(-1)}
                                onLongPress={(): void =>
                                    this.longIncrement(-10)
                                }
                                onPressOut={(): void => this.longIncrement()}
                            >
                                <Ionicons
                                    name={"md-remove"}
                                    size={this.style.iconSize}
                                    color={this.style.textColor}
                                />
                            </TouchableOpacity>
                        </View>
                    )}
                </View>
            </View>
        );
    }

    private setStyle(): void {
        console.debug("currentStyle:", this.style);
        if (!this.style.activeColor) {
            this.style.activeColor = TimePickerDefaultStyle.activeColor;
        }
        if (!this.style.backgroundColor) {
            this.style.backgroundColor = TimePickerDefaultStyle.backgroundColor;
        }
        if (!this.style.borderColor) {
            this.style.borderColor = TimePickerDefaultStyle.borderColor;
        }
        if (!this.style.borderRadius) {
            this.style.borderRadius = TimePickerDefaultStyle.borderRadius;
        }
        if (!this.style.containerBackgroudColor) {
            this.style.containerBackgroudColor =
                TimePickerDefaultStyle.containerBackgroudColor;
        }
        if (!this.style.fontSize) {
            this.style.fontSize = TimePickerDefaultStyle.fontSize;
        }
        if (!this.style.iconSize) {
            this.style.iconSize = TimePickerDefaultStyle.iconSize;
        }
        if (!this.style.textColor) {
            this.style.textColor = TimePickerDefaultStyle.textColor;
        }
    }
    private invokeOnChangeTime = (): void => {
        if (!this.props.onChangeTime) return;
        let h = this.state.hours;
        if (this.state.meridian === "PM" && h !== 12) {
            h = h + 12;
        }
        this.props.onChangeTime(h, this.state.minutes, this.state.seconds);
    };

    private updateHours = (currentHours: number): void => {
        let newHours = currentHours;
        if (this.props.mode24hours === true) {
            newHours = this.calculate24Hours(currentHours);
            this.setState({ hours: newHours });
        } else {
            newHours = this.calculate12Hours(currentHours);
            this.setState({
                hours: newHours,
                meridian: this.getMeridian(newHours)
            });
        }
    };

    private getMeridian = (hours: number): "AM" | "PM" => {
        const meridian = this.state.meridian;
        if (hours > 12) {
            return meridian === "AM" ? "PM" : "AM";
        }
        return meridian;
    };

    private updateMinutes = (minutes: number): void => {
        let tempMinutes = minutes;
        if (minutes > 59) {
            tempMinutes = 0;
            this.updateHours(this.state.hours + 1);
        } else if (minutes < 0) {
            tempMinutes = 59;
            this.updateHours(this.state.hours - 1);
        }
        this.setState({ minutes: tempMinutes });
    };

    private updateSeconds = (seconds: number): void => {
        let tempSeconds = seconds;
        if (seconds > 59) {
            tempSeconds = 0;
            this.updateMinutes(this.state.minutes + 1);
        } else if (seconds < 0) {
            tempSeconds = 59;
            this.updateMinutes(this.state.minutes - 1);
        }
        this.setState({ seconds: tempSeconds });
    };

    private update = (item: Text): void => {
        if (this.state.updating === item) {
            this.setState({ updating: undefined });
            this.updateStyle(item, false);
        } else {
            if (this.state.updating !== undefined) {
                this.updateStyle(this.state.updating, false);
            }
            this.updateStyle(item, true);
            this.setState({ updating: item });
        }
    };

    private increment = (inc: number): void => {
        this.incrementActiveControl(inc);
    };

    private longIncrement = (inc?: number): void => {
        if (inc !== undefined) {
            this.interval = setInterval(() => {
                this.incrementActiveControl(inc);
            }, 500);
        } else {
            clearInterval(this.interval!);
        }
    };

    private calculate12Hours(currentHours: number): number {
        let newHours = currentHours;
        if (currentHours > 12) {
            newHours = 1;
        } else if (currentHours <= 0) {
            newHours = 12;
        }
        return newHours;
    }

    private calculate24Hours(currentHours: number): number {
        let newHours = currentHours;
        if (currentHours > 23) {
            newHours = 0;
        } else if (currentHours < 0) {
            newHours = 23;
        }
        return newHours;
    }

    private incrementActiveControl(inc: number): void {
        if (this.state.updating === this.hoursText) {
            this.updateHours(this.state.hours + inc);
        } else if (this.state.updating === this.minutesText) {
            this.updateMinutes(this.state.minutes + inc);
        } else if (this.state.updating === this.secondsText) {
            this.updateSeconds(this.state.seconds + inc);
        }
    }
    private updateStyle(target: Text, isActive: boolean): void {
        target.setNativeProps({
            style: {
                backgroundColor: isActive
                    ? this.style.activeColor
                    : this.style.backgroundColor,
                borderWidth: isActive ? 2 : 1
            }
        });
    }

    private getContainerColor = (): ViewStyle => {
        return {
            backgroundColor: this.style.containerBackgroudColor
        };
    };

    private getTextColor = (): TextStyle => {
        return { color: this.style.textColor };
    };

    private getTextStyle = (): TextStyle => {
        return {
            fontSize: this.style.fontSize,
            borderRadius: this.style.borderRadius,
            color: this.style.textColor,
            backgroundColor: this.style.backgroundColor,
            borderColor: this.style.borderColor
        };
    };
}

const styles = StyleSheet.create({
    container: {
        margin: 5,
        marginLeft: 7,
        //width: "auto",
        flex: 1,
        height: 70,
        flexDirection: "row",
        justifyContent: "space-around",
        alignItems: "center",
        alignContent: "center",
        borderRadius: 4
    },
    timeContainer: {
        flexDirection: "row",
        margin: 5
    },
    center: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center"
    },
    text: {
        borderWidth: 1,
        paddingHorizontal: 5,
        textAlign: "center",
        justifyContent: "space-around",
        alignItems: "center",
        alignContent: "center"
    },
    increment: {
        marginRight: 5,
        padding: 7
    },
    colon: {
        fontSize: 35,
        marginHorizontal: 3
    },
    meridian: {
        position: "absolute",
        top: 0,
        left: 5
    },
    meridian_text: {
        fontSize: 16
    }
});
