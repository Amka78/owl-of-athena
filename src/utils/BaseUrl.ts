import Constants from "expo-constants";

/**
 * Manage API-URL for each environment.
 *
 * @export
 * @class BaseUrl
 */
export default class BaseUrl {
    public static get(relCh?: string) {
        if (!relCh && Constants.manifest.releaseChannel) {
            relCh = Constants.manifest.releaseChannel;
        }

        if (relCh !== undefined) {
            if (relCh!.includes("staging")) {
                return IWinksRestAPI.staging.url;
            } else if (relCh!.includes("prod")) {
                return IWinksRestAPI.prod.url;
            }
        }

        return IWinksRestAPI.dev.url;
    }

    public static getDownloadUrl(
        image: string | string[],
        index?: number
    ): string | undefined {
        if (typeof image === "undefined") {
            return undefined;
        } else if (typeof image === "string") {
            return BaseUrl.get() + image;
        } else {
            if (index === undefined) {
                index = 0;
            }
            return BaseUrl.get() + image![index];
        }
    }
}

export const IWinksRestAPI = {
    dev: {
        url: "https://api.iwinks.io/"
    },
    staging: {
        url: "https://api.iwinks.io/"
    },
    prod: {
        url: "https://api.iwinks.io/"
    }
};
