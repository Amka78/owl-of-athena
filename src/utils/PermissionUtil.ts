import * as Permissions from "expo-permissions";
import { Linking } from "react-native";

export default class PermissionUtil {
    public static async getPermission(permission: any) {
        const { status } = await Permissions.askAsync(permission);
        if (status !== "granted") {
            Linking.openURL("app-settings:");
            return false;
        }
        return true;
    }
}
