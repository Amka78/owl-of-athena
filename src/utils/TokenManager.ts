import { Platform } from "@unimodules/core";
import * as SecureStore from "expo-secure-store";
import { AsyncStorage } from "react-native";

enum SecureStoreKeys {
    JWT_TOKEN = "jwt_token"
}
/**
 * Token Management Class
 *
 * @export
 * @class TokenManager
 */
export default class TokenManager {
    public static async hasToken(): Promise<boolean> {
        const result = await this.getPerPlatform(SecureStoreKeys.JWT_TOKEN);

        return result !== undefined;
    }

    /**
     * get token
     *
     * @static
     * @returns {(Promise<string | undefined>)}
     * @memberof TokenManager
     */
    public static async get(): Promise<string | undefined> {
        const token = await this.getPerPlatform(SecureStoreKeys.JWT_TOKEN);

        if (token) {
            return token;
        }

        return undefined;
    }

    /**
     * set token
     *
     * @static
     * @param {TokenDto} value
     * @memberof TokenManager
     */
    public static async set(value: string) {
        await this.setPerPlatform(SecureStoreKeys.JWT_TOKEN, value);
    }

    public static async reset() {
        await this.deletePerPlatform(SecureStoreKeys.JWT_TOKEN);
    }

    private static async getPerPlatform(key: string): Promise<string | null> {
        if (Platform.OS === "web") {
            return await AsyncStorage.getItem(key);
        }

        return await SecureStore.getItemAsync(key);
    }

    private static async setPerPlatform(
        key: string,
        value: string
    ): Promise<void> {
        if (Platform.OS === "web") {
            return await AsyncStorage.setItem(key, value);
        }

        return await SecureStore.setItemAsync(key, value);
    }

    private static async deletePerPlatform(key: string): Promise<void> {
        if (Platform.OS === "web") {
            return await AsyncStorage.removeItem(key);
        }

        return await SecureStore.deleteItemAsync(key);
    }
}
