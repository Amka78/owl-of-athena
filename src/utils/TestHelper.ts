import "../components";

import { shallow, ShallowWrapper } from "enzyme";
import enzymetoJson from "enzyme-to-json";
import {
    GestureResponderEvent,
    NativeSyntheticEvent,
    TextInputEndEditingEventData,
} from "react-native";

import { ButtonProps } from "../components/atoms/Button";

// @ts-ignore
export const FakeGestureResponderEvent: GestureResponderEvent = {};
// @ts-ignore
export const FakeTextBoxOnEndEditingEventData: NativeSyntheticEvent<TextInputEndEditingEventData> = {};

export function createMock<P, S>(comp: any): ShallowWrapper<P, S> {
    return shallow<P, S>(comp);
}

export function toJson(wrapper: any) {
    return enzymetoJson(wrapper);
}

/**
 * TestIDを利用してComponetのPropsを取得します。
 *
 * @static
 * @template T Propsの型を指定
 * @param {ShallowWrapper} component
 * @param {string} name
 * @param {string} testId
 * @returns {T}
 * @memberof TestHelper
 */
export function getComponentPropsByTestID<T>(
    component: ShallowWrapper<any, any>,
    name: string,
    testId: string
): T | undefined {
    const findComponent = component.find(name).findWhere((wrapper: any) => {
        const targetProps = wrapper.props();
        return targetProps.testID === testId;
    });
    if (findComponent.length <= 0) {
        return undefined;
    }
    const props: T = findComponent.props();
    return props;
}

export function getButtonPropsByTestID(
    component: ShallowWrapper<any, any>,
    testID: string
): ButtonProps | undefined {
    return getComponentPropsByTestID<ButtonProps>(
        component,
        "MultiLingualButton",
        testID
    );
}

/*public static getTextBoxPropsByTestID(
        component: ShallowWrapper<any, any>,
        testID: string
    ): ITextBoxProps | undefined {
        return this.getComponentPropsByTestID<ITextBoxProps>(
            component,
            "MultiLingualTextBox",
            testID
        );
    }

    public static getHeaderSearchBarPropsByTestID(
        component: ShallowWrapper<any, any>,
        testID: string
    ): IHeaderSearchBarProps | undefined {
        return this.getComponentPropsByTestID<IHeaderSearchBarProps>(
            component,
            "HeaderSearchBar",
            testID
        );
    }

    public static getListItemPropsByTestID(
        component: ShallowWrapper<any, any>,
        testID: string
    ): IListItemProps | undefined {
        return this.getComponentPropsByTestID<IListItemProps>(
            component,
            "MultiLingualListItem",
            testID
        );
    }*/

/*public static createNavigationMock(
        param: Partial<NavigationScreenProp<any, any>>
    ): NavigationScreenProp<any, any> {
        const navigation = new Object();
        Object.assign(navigation, param);
        return navigation as NavigationScreenProp<any, any>;
    }*/
