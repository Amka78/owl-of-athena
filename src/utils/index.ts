import BaseUrl from "./BaseUrl";
export { BaseUrl };
