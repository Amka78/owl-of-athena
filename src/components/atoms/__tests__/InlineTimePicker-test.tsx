//#region Import Modules
import "react-native";

import { ShallowWrapper } from "enzyme";
import React from "react";

import { createMock, toJson } from "../../../utils/TestHelper";
import { InlineTimePicker, InlineTimePickerProps } from "../InlineTimePicker";
//#endregion

//#region Test
let component: ShallowWrapper<InlineTimePickerProps, unknown, unknown>;
describe("InlineTimePicker UnitTest", () => {
    it("Renders correctly", () => {
        component = createMock(<InlineTimePicker></InlineTimePicker>);

        expect(toJson(component)).toMatchSnapshot();
    });
});
//#endregion
