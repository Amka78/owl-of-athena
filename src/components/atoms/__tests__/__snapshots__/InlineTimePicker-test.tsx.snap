// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineTimePicker UnitTest Renders correctly 1`] = `
<InlineTimePickerCore
  theme={
    Object {
      "animation": Object {
        "scale": 1,
      },
      "colors": Object {
        "accent": "#03dac4",
        "backdrop": "rgba(0, 0, 0, 0.5)",
        "background": "#f6f6f6",
        "disabled": "rgba(0, 0, 0, 0.26)",
        "error": "#B00020",
        "notification": "#f50057",
        "onBackground": "#000000",
        "onSurface": "#000000",
        "placeholder": "rgba(0, 0, 0, 0.54)",
        "primary": "#6200ee",
        "surface": "#ffffff",
        "text": "#000000",
      },
      "dark": false,
      "fonts": Object {
        "light": Object {
          "fontFamily": "System",
          "fontWeight": "300",
        },
        "medium": Object {
          "fontFamily": "System",
          "fontWeight": "500",
        },
        "regular": Object {
          "fontFamily": "System",
          "fontWeight": "400",
        },
        "thin": Object {
          "fontFamily": "System",
          "fontWeight": "100",
        },
      },
      "roundness": 4,
    }
  }
/>
`;
